### PR TITLE
Add ability to read and write from LocalStorage

### PR DESCRIPTION
**Pull request by abidlabs**
I've added a component `gr.BrowserState()`, which acts similarly to `gr.State()` but any value written to it is saved to the browser's local storage, so it is preserved across page loads. The value is encrypted with a key, which means that one Gradio app will not be able to read the local state value of another Gradio app. 

Test demo: `demo/browserstate/run.py`

Closes: https://github.com/gradio-app/gradio/issues/3106

### CODE REVIEW

**Review by julien-c — COMMENTED** [review]
cool feature :fire:

**Comment by abidlabs** [comment]
I think so, otherwise the frontend doesn’t recognize the existence of the component

**Comment by aliabid94** [comment]
The idea of this is cool, I think even beyond the persistence, users may want access to a State variable that can be accessed through custom js. They actually may not want this to even "persist", but specifically only just want the feature that it is accessible via js.
I am not sure if the name `gr.LocalState` is clear about what it does. Maybe `gr.BrowserState`?
I have issues with this implementation, I believe all the code to handle storage and retrieval can actually be stored in `js/localstate/Index.svelte` and that's the correct place to implement it, not `Blocks.svelte`. Let's not store component specific logic there. Lmk if you want to sync on how we can implement it in the component code.

**Comment by abidlabs** [comment]
> I am not sure if the name gr.LocalState is clear about what it does. Maybe gr.BrowserState?

Agreed, will rename it to `gr.BrowserState`

> I have issues with this implementation, I believe all the code to handle storage and retrieval can actually be stored in js/localstate/Index.svelte and that's the correct place to implement it, not Blocks.svelte. Let's not store component specific logic there. Lmk if you want to sync on how we can implement it in the component code.

Ah yeah that makes sense, I'll move the logic over to `js/localstate/Index.svelte`

> The idea of this is cool, I think even beyond the persistence, users may want access to a State variable that can be accessed through custom js. They actually may not want this to even "persist", but specifically only just want the feature that it is accessible via js. I think it may be a good idea to make the persistence configurable.

I think this is a separate feature, and should _already_ be doable with the `head` parameter. I'll keep this PR focused on browser state / localstorage

**Comment by abidlabs** [comment]
Should be ready for review now, thanks @aliabid94 for helping me sort out a few Svelte things!

**Review by aliabid94 — APPROVED** [review]
LGTM thanks for fixing the issues!

File: js/app/src/routes/[...catchall]/+page.svelte [hunk 1]
```diff
@@ -30,6 +30,7 @@
 		stylesheets?: string[];
 		path: string;
 		app_id?: string;
+		browser_state_secret?: string;
```
Comment (aliabid94) [inline js/app/src/routes/[...catchall]/+page.svelte:33]: is this still used?

File: js/spa/src/Index.svelte [hunk 2]
```diff
@@ -29,6 +29,7 @@
 		stylesheets?: string[];
 		path: string;
 		app_id?: string;
+		browser_state_secret?: string;
```
Comment (aliabid94) [inline js/spa/src/Index.svelte:32]: and this one?

**Review by aliabid94 — APPROVED** [review]
LGTM good stuff!

**Comment by abidlabs** [comment]
Thanks @aliabid94!

**Comment by ROBERT-MCDOWELL** [comment]
will it be in the next release version?

**Comment by abidlabs** [comment]
@ROBERT-MCDOWELL yes which we are planning for today / Monday

**Comment by ROBERT-MCDOWELL** [comment]
ok good to know as I had impletemented it with my own solution a week ago.....

**Pull request merged**